### DISCUSSION TITLE
Any way to know if memcpy is submitted into compute engine or copy engine?

### DISCUSSION THREAD

**Discussion by GHGmc2** [intel/compute-runtime#430] [Q&A]
Is there any way to know if memcpy is submitted into compute engine or copy engine? 
Or more generally, any env variables or logs on which command queue are command lists submitted to?

**Comment by jandres742** [comment]
For OpenCL, copy operations use the copy engine if available.
For Level Zero, application needs to use queue groups to select the copy engine. See here https://spec.oneapi.com/level-zero/latest/core/PROG.html#command-queue-groups

For OpenCL, you can use the following debug keys to steer behavior:

EnableBlitterOperationsSupport
EnableBlitterForEnqueueOperations
EnableBlitterForReadWriteImage

https://github.com/intel/compute-runtime/blob/233aab88e707e6a314ddf0489ca1c6238eaeafd8/shared/source/debug_settings/debug_variables_base.inl#L250

**Reply by GHGmc2** [reply]
Thanks!
On application(framework) side, we do not call Level Zero APIs directly, but through SYCL APIs and DPC++ compiler implementations, any way(with env variables or other ways) to know if we are using copy engine on application side?

**Reply by MichalMrozek** [reply]
You may also use this variable -> PrintDeviceAndEngineIdOnSubmission it prints following details:
- to what device driver submits
-  what is the engine that is getting submission ( compute / copy , etc.)

**Reply by GHGmc2** [reply]
Thanks, I'll have a try.

**Reply by jandres742** [reply]
For SYCL APIs when using L0, you need to use `SYCL_PI_LEVEL_ZERO_USE_COPY_ENGINE=1`

https://github.com/intel/llvm/blob/06e4ebc73bc9093d2a2b327598833a787b51d2a0/sycl/plugins/level_zero/pi_level_zero.cpp#L548

For SYCL over OpenCL, same rules mentioned before applied. Also, you may use debug key shared by @MichalMrozek
